### PR TITLE
Adds the cancellable ban events and implements them in the server, resolves #678

### LINKED ISSUE #678: please, Add cancellable (un)ban(ip) Events
Hello. I think everything is clear from the title.
I don't understand why this is not done in a Spigot and I will be glad if at least you make it.

@Zbob750, so what do you think about my proposal to add beautiful explosions?

### PR DESCRIPTION
**Pull request by dignissimus**

### CODE REVIEW

File: Spigot-API-Patches/0067-Added-ban-events-resolves-678.patch [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+From: Sam Ezeh <bubuezeh@gmail.com>
+Date: Wed, 30 Aug 2017 21:32:18 +0100
+Subject: [PATCH] Added ban events,  resolves #678
+
+
+diff --git a/src/main/java/org/bukkit/event/server/BanEvent.java b/src/main/java/org/bukkit/event/server/BanEvent.java
+new file mode 100644
+index 00000000..dcd95bf6
+--- /dev/null
++++ b/src/main/java/org/bukkit/event/server/BanEvent.java
```
Comment (electronicboy) [inline Spigot-API-Patches/0067-Added-ban-events-resolves-678.patch]: if there is anything we have learnt, any classes we add should ideally be in our own package as opposed to bukkits

File: Spigot-Server-Patches/0236-Implemented-the-cancellable-ban-events.patch [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+
+
+diff --git a/src/main/java/org/bukkit/craftbukkit/CraftIpBanList.java b/src/main/java/org/bukkit/craftbukkit/CraftIpBanList.java
+index 80832f78..05791767 100644
+--- a/src/main/java/org/bukkit/craftbukkit/CraftIpBanList.java
++++ b/src/main/java/org/bukkit/craftbukkit/CraftIpBanList.java
+@@ -1,19 +1,19 @@
+ package org.bukkit.craftbukkit;
+ 
+-import java.io.IOException;
```
Comment (electronicboy) [inline Spigot-Server-Patches/0236-Implemented-the-cancellable-ban-events.patch:154]: Don't reorganise imports, you might wanna disable this feature in your IDE for dealing with paper stuff as it causes issues like this that make handling diffs a headache for everyone.

File: Spigot-Server-Patches/0236-Implemented-the-cancellable-ban-events.patch [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+     public org.bukkit.BanEntry addBan(String target, String reason, Date expires, String source) {
++
+         Validate.notNull(target, "Ban target cannot be null");
+ 
+         IpBanEntry entry = new IpBanEntry(target, new Date(),
+                 StringUtils.isBlank(source) ? null : source, expires,
+                 StringUtils.isBlank(reason) ? null : reason);
+ 
+-        list.add(entry);
++        BanIpEvent event = new BanIpEvent(new CraftIpBanEntry(target, entry, list));
```
Comment (electronicboy) [inline Spigot-Server-Patches/0236-Implemented-the-cancellable-ban-events.patch:191]: // paper (more places than just this)

File: Spigot-Server-Patches/0236-Implemented-the-cancellable-ban-events.patch [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
++        event = CraftEventFactory.callEvent(event);
++
++        if(event.isCancelled()) return null;
```
Comment (electronicboy) [inline Spigot-Server-Patches/0236-Implemented-the-cancellable-ban-events.patch:194]: This API is pretty much contracted to not return null, to change that contract now would potentially be breaking

File: Spigot-Server-Patches/0236-Implemented-the-cancellable-ban-events.patch [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+ public class CraftProfileBanList implements org.bukkit.BanList {
+     private final GameProfileBanList list;
+@@ -54,7 +54,14 @@ public class CraftProfileBanList implements org.bukkit.BanList {
+                 StringUtils.isBlank(source) ? null : source, expires,
+                 StringUtils.isBlank(reason) ? null : reason);
+ 
+-        list.add(entry);
++        BanProfileEvent event = new BanProfileEvent(new CraftProfileBanEntry(profile, entry, list));
++        event = CraftEventFactory.callEvent(event);
++        if(event.isCancelled()) return null;
```
Comment (electronicboy) [inline Spigot-Server-Patches/0236-Implemented-the-cancellable-ban-events.patch:251]: null

File: Spigot-API-Patches/0067-Added-ban-events-resolves-678.patch [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
++/**
++ * Called when a player is banned
++ */
++public abstract class BanEvent extends ServerEvent implements Cancellable{
++    private final Date created;
++    private Date expiration;
++    private String source;
++    private String reason;
++    private String target;
++    private boolean cancel = true;
```
Comment (electronicboy) [inline Spigot-API-Patches/0067-Added-ban-events-resolves-678.patch]: true?

File: Spigot-Server-Patches/0236-Implemented-the-cancellable-ban-events.patch (hunk 5 above)
Comment (electronicboy) [inline Spigot-Server-Patches/0236-Implemented-the-cancellable-ban-events.patch:251]: I am partially concerned with the practices in this PR, this really isn't all that inclusive either due to how CB has wrapped this, in the event that a ban is done with the vanilla command, this won't actually be caught by this PR.

I would say that these changes are actually better being made in the NMS BanList's, that way it's more inclusive

**Review by electronicboy — CHANGES_REQUESTED** [review]

**Comment by dignissimus** [comment]
Changing source branch

**Pull request closed without merging**